pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -1052,6 +1059,55 @@ func pruneBundle(tx *gorm.DB, trustDomainID string, expiry time.Time, log logrus
 	return changed, nil
 }
 
+func taintJWTKey(tx *gorm.DB, trustDomainID string, taintedKeyID string) (*common.PublicKey, error) {
+	model := &Bundle{}
+	if err := tx.Find(model, "trust_domain = ?", trustDomainID).Error; err != nil {
+		return nil, sqlError.Wrap(err)
+	}
+
+	bundle, err := modelToBundle(model)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to unmarshal bundle: %v", err)
+	}
+
+	var taintedKey *common.PublicKey
+	for _, jwtKey := range bundle.JwtSigningKeys {
+		if jwtKey.Kid != taintedKeyID {
+			continue
+		}
+
+		if jwtKey.TaintedKey {
+			return nil, status.Error(codes.Internal, "key is already tainted")
+		}
+
+		// Breaking here to be sure that we have only one
+		// key with the giver KeyID, it must never happens
+		// since we does not allow to have repeated key IDs
```

Comment:
Maybe this could be rephrased:

```suggestion
		// Check if a JWT Key with the provided keyID was already
		// tainted in this loop. This is purely defensive since we do not
		// allow to have repeated key IDs.
```